### ISSUE TITLE
Same decision id maps to different request on client vs server side

### ISSUE THREAD

**Issue by xhl1988** [open-policy-agent/opa#5720]
## Short description 

OPA version: 0.30.1, 0.49.2

For v0.30.1
Client log:
```
2023-03-02 23:32:53,025 WARN Querying OPA with object: {"input":{"requestContext":{"clientAddress":"/1.2.3.4","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"name":{"clientId":"internal-logstash-0","correlationId":430387,"requestApiKey":1,"requestApiVersion":12},"headerVersion":2},"listenerName":"TLS-9093","principal":{"principalType":"User","name":"CN=logstash"},"securityProtocol":"SSL"},"action":{"resourcePattern":{"resourceType":"TOPIC","name":"service-log","patternType":"LITERAL","unknown":false},"operation":"READ","resourceReferenceCount":1,"logIfAllowed":true,"logIfDenied":true}}},
Response code: 200, body: {"decision_id":"01db5070-1f0c-4539-b1b6-d92a178057cd","result":false} (org.openpolicyagent.kafka.AllowCallable) [pool-9-thread-8]
```
server log:
```
{"decision_id":"01db5070-1f0c-4539-b1b6-d92a178057cd","input":{"action":{"logIfAllowed":true,"logIfDenied":false,"operation":"DESCRIBE","resourcePattern":{"name":"__strimzi_store_topic","patternType":"LITERAL","resourceType":"TOPIC","unknown":false},"resourceReferenceCount":1},"requestContext":{"clientAddress":"/1.2.3.4","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"headerVersion":2,"name":{"clientId":"internal-logstash-0","correlationId":384532,"requestApiKey":3,"requestApiVersion":11}},"listenerName":"TLS-9093","principal":{"name":"CN=logstash","principalType":"User"},"securityProtocol":"SSL"}},"labels":{"id":"972d6977-3aa4-4198-bdb0-51425f1819f3","version":"0.30.1"},"level":"info","metrics":{"counter_server_query_cache_hit":1,"timer_rego_external_resolve_ns":1941,"timer_rego_input_parse_ns":74134,"timer_rego_query_eval_ns":99477456,"timer_server_handler_ns":99596591},"msg":"Decision Log","path":"kafka/authz/prod/v0_27_1/allow","requested_by":"172.24.139.226:54332","result":false,"time":"2023-03-02T23:32:53Z","timestamp":"2023-03-02T23:32:53.002983617Z","type":"openpolicyagent.org/decision_logs"}
{"client_addr":"xxx","level":"info","msg":"Sent response.","req_id":8030,"req_method":"POST","req_path":"/v1/data/kafka/authz/prod/v0_27_1/allow","resp_body":"{\"decision_id\":\"01db5070-1f0c-4539-b1b6-d92a178057cd\",\"result\":false}","resp_bytes":69,"resp_duration":99.941556,"resp_status":200,"time":"2023-03-02T23:32:53Z"}
```

For 0.49.2:
client log:
```
2023-03-03 04:58:49,873 WARN Querying OPA with object: {"input":{"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"name":{"clientId":"internal-logstash-0","correlationId":715528,"requestApiKey":1,"requestApiVersion":12},"headerVersion":2},"listenerName":"TLS-9093","principal":{"principalType":"User","name":"CN=logstash"},"securityProtocol":"SSL"},"action":{"resourcePattern":{"resourceType":"TOPIC","name":"service-log","patternType":"LITERAL","unknown":false},"operation":"READ","resourceReferenceCount":1,"logIfAllowed":true,"logIfDenied":true}}},                                 Response code: 200, body: {"decision_id":"729aa9ae-a4e0-41cb-8e1e-958403839820","result":false}
 (org.openpolicyagent.kafka.AllowCallable) [pool-9-thread-5]
```

server log:
```
{"decision_id":"729aa9ae-a4e0-41cb-8e1e-958403839820","input":{"action":{"logIfAllowed":true,"logIfDenied":false,"operation":"CLUSTER_ACTION","resourcePattern":{"name":"kafka-cluster","patternType":"LITERAL","resourceType":"CLUSTER","unknown":false},"resourceReferenceCount":1},"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"headerVersion":2,"name":{"clientId":"internal-logstash-0","correlationId":592953,"requestApiKey":23,"requestApiVersion":4}},"listenerName":"TLS-9093","principal":{"name":"CN=logstash","principalType":"User"},"securityProtocol":"SSL"}},"labels":{"id":"e3213400-62e6-49a9-86ea-e539eb545e36","version":"0.49.2"},"level":"info","metrics":{"counter_server_query_cache_hit":1,"timer_rego_external_resolve_ns":331,"timer_rego_input_parse_ns":54991,"timer_rego_query_eval_ns":418573,"timer_server_handler_ns":494302},"msg":"Decision Log","path":"kafka/authz/prod/v0_27_1/allow","req_id":1270,"requested_by":"xxx","result":false,"time":"2023-03-03T04:58:49Z","timestamp":"2023-03-03T04:58:49.86061984Z","type":"openpolicyagent.org/decision_logs"}
{"client_addr":"xxx","level":"info","msg":"Sent response.","req_id":1270,"req_method":"POST","req_path":"/v1/data/kafka/authz/prod/v0_27_1/allow","resp_body":"{\"decision_id\":\"729aa9ae-a4e0-41cb-8e1e-958403839820\",\"result\":false}\n","resp_bytes":70,"resp_duration":0.731548,"resp_status":200,"time":"2023-03-03T04:58:49Z"}
```

## Steps To Reproduce

We use https://github.com/StyraInc/opa-kafka-plugin with Strimzi, it constantly happens in all deployments.

**Comment by anderseknert** [comment]
Hi there! And thanks for reporting this. As the decision ID is created by OPA (ie on the server) this must be an issue in the Kafka plug-in. Could you please close this here and re-open in that repo?

**Comment by xhl1988** [comment]
@anderseknert here is the opa plugin code related to the opa call:
```
  override def call(): Boolean = {
    logger.debug(s"Cache miss, querying OPA for decision")
    val reqJson = AllowCallable.objectMapper.writeValueAsString(request)
    val req = AllowCallable.requestBuilder.uri(opaUrl).POST(BodyPublishers.ofString(reqJson)).build
    try {
      val resp = httpClient.send(req, BodyHandlers.ofString)
      logger.warn(s"Querying OPA with object: $reqJson,                                 Response code: ${resp.statusCode}, body: ${resp.body}")
      val result = AllowCallable.objectMapper.readTree(resp.body()).at("/result").asBoolean
      result
    } catch {
      case e: Exception =>
        logger.warn(s"Exception when calling OPA server with request: $request")
        if (metrics.isDefined) {
          metrics.get.sensor(MetricsLabel.REQUEST_TO_OPA_ERROR_COUNT).record()
        }
        throw e
    } finally {
      if (metrics.isDefined) {
        metrics.get.sensor(MetricsLabel.REQUEST_TO_OPA_LATENCY).record((System.currentTimeMillis() - createTime).toDouble)
      }
    }
  }
```
The `decision_id` is got from the `resp.body`, I think it should be returned from the opa server?

**Comment by xhl1988** [comment]
Here is the most recent logs:
From client side, three consecutive logs, all `logstash` to `service-log-kafka`, first one `DESCRIBE` succeeded, second one `READ` succeeded, third one `DESCRIBE` failed:
```
2023-03-03 06:05:31,579 WARN Querying OPA with object: {"input":{"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"name":{"clientId":"internal-logstash-0","correlationId":10953,"requestApiKey":3,"requestApiVersion":11},"headerVersion":2},"listenerName":"TLS-9093","principal":{"principalType":"User","name":"CN=logstash"},"securityProtocol":"SSL"},"action":{"resourcePattern":{"resourceType":"TOPIC","name":"service-log-kafka","patternType":"LITERAL","unknown":false},"operation":"DESCRIBE","resourceReferenceCount":1,"logIfAllowed":true,"logIfDenied":false}}},                                 Response code: 200, body: {"decision_id":"11c9626e-59b9-467b-95fd-a072c7f5fab6","result":true}
2023-03-03 06:05:35,268 WARN Querying OPA with object: {"input":{"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"name":{"clientId":"internal-logstash-0","correlationId":1981811,"requestApiKey":1,"requestApiVersion":12},"headerVersion":2},"listenerName":"TLS-9093","principal":{"principalType":"User","name":"CN=logstash"},"securityProtocol":"SSL"},"action":{"resourcePattern":{"resourceType":"TOPIC","name":"service-log-kafka","patternType":"LITERAL","unknown":false},"operation":"READ","resourceReferenceCount":1,"logIfAllowed":true,"logIfDenied":true}}},                                 Response code: 200, body: {"decision_id":"c51f78b8-b859-49b8-9b56-1b20c85daa63","result":true}
2023-03-03 06:05:35,269 WARN Querying OPA with object: {"input":{"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"name":{"clientId":"internal-logstash-0","correlationId":1981809,"requestApiKey":23,"requestApiVersion":4},"headerVersion":2},"listenerName":"TLS-9093","principal":{"principalType":"User","name":"CN=logstash"},"securityProtocol":"SSL"},"action":{"resourcePattern":{"resourceType":"TOPIC","name":"service-log-kafka","patternType":"LITERAL","unknown":false},"operation":"DESCRIBE","resourceReferenceCount":1,"logIfAllowed":true,"logIfDenied":true}}},                                 Response code: 200, body: {"decision_id":"0eb0a03c-177f-45f0-9b08-7bfcfe036460","result":false}
```

Using `decision_id` to find on server side:
```
{"decision_id":"11c9626e-59b9-467b-95fd-a072c7f5fab6","input":{"action":{"logIfAllowed":true,"logIfDenied":false,"operation":"DESCRIBE","resourcePattern":{"name":"service-log-kafka","patternType":"LITERAL","resourceType":"TOPIC","unknown":false},"resourceReferenceCount":1},"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"headerVersion":2,"name":{"clientId":"internal-logstash-0","correlationId":10953,"requestApiKey":3,"requestApiVersion":11}},"listenerName":"TLS-9093","principal":{"name":"CN=logstash","principalType":"User"},"securityProtocol":"SSL"}},"labels":{"id":"30984a11-6cf5-454b-b830-a16be6eb6ece","version":"0.49.2"},"level":"info","metrics":{"counter_server_query_cache_hit":1,"timer_rego_external_resolve_ns":1000,"timer_rego_input_parse_ns":111401,"timer_rego_query_eval_ns":833606,"timer_server_handler_ns":969607},"msg":"Decision Log","path":"kafka/authz/prod/v0_27_1/allow","req_id":1855,"requested_by":"xxx","result":true,"time":"2023-03-03T06:05:31Z","timestamp":"2023-03-03T06:05:31.55459396Z","type":"openpolicyagent.org/decision_logs"}
{"client_addr":"xxx","level":"info","msg":"Sent response.","req_id":1855,"req_method":"POST","req_path":"/v1/data/kafka/authz/prod/v0_27_1/allow","resp_body":"{\"decision_id\":\"11c9626e-59b9-467b-95fd-a072c7f5fab6\",\"result\":true}\n","resp_bytes":69,"resp_duration":1.739811,"resp_status":200,"time":"2023-03-03T06:05:31Z"}

{"decision_id":"c51f78b8-b859-49b8-9b56-1b20c85daa63","input":{"action":{"logIfAllowed":true,"logIfDenied":true,"operation":"READ","resourcePattern":{"name":"service-log-kafka","patternType":"LITERAL","resourceType":"TOPIC","unknown":false},"resourceReferenceCount":1},"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"headerVersion":2,"name":{"clientId":"internal-logstash-0","correlationId":1981811,"requestApiKey":1,"requestApiVersion":12}},"listenerName":"TLS-9093","principal":{"name":"CN=logstash","principalType":"User"},"securityProtocol":"SSL"}},"labels":{"id":"30984a11-6cf5-454b-b830-a16be6eb6ece","version":"0.49.2"},"level":"info","metrics":{"counter_server_query_cache_hit":1,"timer_rego_external_resolve_ns":800,"timer_rego_input_parse_ns":83901,"timer_rego_query_eval_ns":544103,"timer_server_handler_ns":653804},"msg":"Decision Log","path":"kafka/authz/prod/v0_27_1/allow","req_id":1940,"requested_by":"xxx","result":true,"time":"2023-03-03T06:05:35Z","timestamp":"2023-03-03T06:05:35.264486601Z","type":"openpolicyagent.org/decision_logs"}
{"client_addr":"xxx","level":"info","msg":"Sent response.","req_id":1940,"req_method":"POST","req_path":"/v1/data/kafka/authz/prod/v0_27_1/allow","resp_body":"{\"decision_id\":\"c51f78b8-b859-49b8-9b56-1b20c85daa63\",\"result\":true}\n","resp_bytes":69,"resp_duration":0.929806,"resp_status":200,"time":"2023-03-03T06:05:35Z"}

{"decision_id":"0eb0a03c-177f-45f0-9b08-7bfcfe036460","input":{"action":{"logIfAllowed":true,"logIfDenied":false,"operation":"CLUSTER_ACTION","resourcePattern":{"name":"kafka-cluster","patternType":"LITERAL","resourceType":"CLUSTER","unknown":false},"resourceReferenceCount":1},"requestContext":{"clientAddress":"/xxx","clientInformation":{"softwareName":"apache-kafka-java","softwareVersion":"2.8.1"},"connectionId":"xxx","header":{"headerVersion":2,"name":{"clientId":"internal-logstash-0","correlationId":1981809,"requestApiKey":23,"requestApiVersion":4}},"listenerName":"TLS-9093","principal":{"name":"CN=logstash","principalType":"User"},"securityProtocol":"SSL"}},"labels":{"id":"30984a11-6cf5-454b-b830-a16be6eb6ece","version":"0.49.2"},"level":"info","metrics":{"counter_server_query_cache_hit":1,"timer_rego_external_resolve_ns":400,"timer_rego_input_parse_ns":78001,"timer_rego_query_eval_ns":476003,"timer_server_handler_ns":584104},"msg":"Decision Log","path":"kafka/authz/prod/v0_27_1/allow","req_id":1937,"requested_by":"xxx","result":false,"time":"2023-03-03T06:05:35Z","timestamp":"2023-03-03T06:05:35.26122498Z","type":"openpolicyagent.org/decision_logs"}
{"client_addr":"xxx","level":"info","msg":"Sent response.","req_id":1937,"req_method":"POST","req_path":"/v1/data/kafka/authz/prod/v0_27_1/allow","resp_body":"{\"decision_id\":\"0eb0a03c-177f-45f0-9b08-7bfcfe036460\",\"result\":false}\n","resp_bytes":70,"resp_duration":0.975207,"resp_status":200,"time":"2023-03-03T06:05:35Z"}
```

The third one becomes `CLUSTER_ACTION` instead of `DESCRIBE`.

**Comment by anderseknert** [comment]
> The decision_id is got from the resp.body, I think it should be returned from the opa server?

Right, and the expectation is that the request logged as sent in the client should be the same logged as received in OPA. That makes me think something odds is going on in the client when logging the request. I don't see anything immediately obvious in the code though. I'd still try and Occam this by removing the Kafka plugin from the equation. If you send a bunch of requests using curl, or the HTTP client in your scripting language of choice, do you ever see OPA logging a different request than the one sent?

I'll leave this open here until we've figured out where the error is caused.

CC @scholzj

**Comment by xhl1988** [comment]
This is how we construct the client:
```
// keyManagerFactory staff
// trustManagerFactory staff
HttpClient.newBuilder.sslContext(sslContext).connectTimeout(ofSeconds(5)).build
```
We can give the `curl` a try, but this is a standard httpClient with a blocking call. 
```
val resp = httpClient.send(req, BodyHandlers.ofString)
logger.warn(s"Querying OPA with object: $reqJson, Response code: ${resp.statusCode}, body: ${resp.body}")
```
Any known issue to that?

**Comment by anderseknert** [comment]
No known issue until possibly now. OPA however logs millions of decisions per minute in all kinds of systems around the globe, while the OPA Kafka plug-in isn't nearly as battle-tested. Hence why I suspect the problem is more likely to be found there than in OPA, and would want to exclude that from the equation. I'm not ruling out an issue in OPA, but if there is one it should manifest regardless of client.

**Comment by scholzj** [comment]
TBH, this is a bit confusing. The source code and logs do not look like the code in https://github.com/StyraInc/opa-kafka-plugin ... and it's not just the one log message which changed. That makes it hard to reproduce if the code is basically custom. It would be also good to understand the configuration of the plugin, possibly also the Strimzi version (as Java versions would differ) etc. The code in the original repo does not seem to suggest any problem TBH

I think the idea to try it with `curl` is good one - it should be quite simple to check and you can run it in parallel with your original Kafka cluster to have some load there at the same time.

**Comment by scholzj** [comment]
One more thing -> the environment is a bit unclear to me. How many Kafka clusters are authorizing against the given OPA server? How many logstash clients are running in parallel? Not having a full log makes a lot of it a bit unclear.

But in general, the 3rd request which is DESCRIBE TOPIC in Kafka log and CLUSTER_ACTION in the decision log -> they are likely to come from the same `OffsetForLeaderEpoch` Kafka API request which has dual authorization:
* It first checks the `CLUSTER_ACTION` ACL
* And if it fails, it checks the `DESCRIBE` `TOPIC` ACL

And these calls would be normally done sequentially: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L2623-L2629

**Comment by anderseknert** [comment]
Thanks, @scholzj! As always, insightful. I had not realized the plug-in code had been modified. Unless this is reproduced in the plug-in bundled with Strimzi, or another client, like `curl`... I don't think there's much suggesting that this would be a problem in OPA.

**Comment by scholzj** [comment]
> I had not realized the plug-in code had been modified. Unless this is reproduced in the plug-in bundled with Strimzi, or another client

To be clear, in the code snippet shared above, I do not see anything that I think would cause this. But it is not clear what is changed besides it, how was it built, etc. That said, testing it with the original code might not be easy as it does not have a single log for the request and response. So when you have some load on the broker, you will have issues pairing the request and the response as they will be logged separately. That is why I think `curl` would be useful here.

**Comment by xhl1988** [comment]
Thanks @anderseknert and @scholzj  for all the insights and suggestion!

TLDR: We have tried the `curl` approach while keeping the normal traffic from Kafka broker, we didn't get a wrong result from `curl`

Context on why we modify the plugin: 1. we used to see denied request from time to time (not that frequent, but enough for us to do something); 2. we saw issues when opa server was restarted due to our internal infra (some routing issue). So we decide to change the sync loading to async loading/refresh to improve the hit rate, and the above problem happens after that.

What we did since last comment:
1. Using the curl, and got no wrong result
2. Tried `expireAfterWrite` and `refreshAfterWrite` with same interval (5min), `expireAfterWrite` didn't give the wrong result while `refreshAfterWrite` can constantly give the wrong result. Code:
```
  private lazy val executor = new ThreadPoolExecutor(20, 200,  60L, TimeUnit.SECONDS, new LinkedBlockingQueue[Runnable]())
  private lazy val cache = CacheBuilder.newBuilder
    .initialCapacity(1000)
    .maximumSize(maxCacheCapacity)
    .expireAfterWrite(86400, TimeUnit.SECONDS)
    .refreshAfterWrite(300, TimeUnit.SECONDS)
    .recordStats().ticker(ticker)
    .build(
      CacheLoader.asyncReloading(
        new CacheLoader[CacheableRequest, Boolean]() {
          override def load(k: CacheableRequest): Boolean = {
            val request = Request(Input(k.requestContext, k.action))
            val callable = new AllowCallable(client, request, opaUrl, allowOnError, metrics)
            val result = callable.call()
            result
          }
        }, executor)
    )
    .asInstanceOf[LoadingCache[CacheableRequest, Boolean]]
```

With 1 & 2, we thought there was something wrong the our async loading logic, but could find out what, so we tried:
3. Change the executor from above to `Executors.newFixedThreadPool(2000)`, and the chance of wrong result was somehow lower than above or `Executors.newFixedThreadPool(200)`
4. Tried `ForkJoinPool.commonPool()` also same issue
5. Tried `Caffeine` cache with both `expireAfterWrite` and `refreshAfterWrite`, same result that `refreshAfterWrite` constantly had the same issue.

Please advice how to investigate this further?

**Comment by anderseknert** [comment]
Hi again @xhl1988 👋 Except for
> 1. we used to see denied request from time to time (not that frequent, but enough for us to do something)

Which I'd be happy to see an issue opened for in the OPA Kafka plug-in repo, I don't think there's much the developers of OPA, or the OPA Kafka plug-in, can do to support your custom modifications. Perhaps try the OPA Slack, but as this seems more of an issue in how the cache is configured or queried, maybe a Java/JVM community channel would be a better target? Closing this here as AFAICS, this isn't an issue in OPA.
